### PR TITLE
HTS concurrent put and delete resolution

### PR DESCRIPTION
**Pull request by jiang95-dev**
## Summary

This PR provides a resolution for the concurrent put and delete bug in HTS. It also fixes the concurrent posts issue.

## Changes

- [ ] Client-facing API Changes
- [ ] Internal API Changes
- [x] Bug Fixes
- [ ] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [ ] Refactoring
- [ ] Documentation
- [ ] Tests

Added `INITIAL_VERSION` check to prevent a put request being blindly accepted when the entry has been deleted. Also changed the initial version of a row to be null so that it can be considered as a new entry to JPA.

## Testing Done

- [x] Manually Tested on local docker setup. Please include commands ran, and their output.
- [ ] Added new tests for the changes made.
- [ ] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

Added temporary code to facilitate tests: Put thread will sleep 5s before calling the save() method.

There are 4 sets of docker tests carried out before and after the change:
### Before change
**Ex1. concurrent posts**
Succeeded. Second post will be taken as a put.

**Ex2. concurrent puts**
Failed. Second put will conflict.

**Ex3. concurrent put and delete**
Failed. Delete goes before save(), and so put will conflict.
```
{"status":"CONFLICT","error":"Conflict","message":"Entity with key[databaseId : d1, tableId : t1, version: 1 The requested user table has been modified/created by other processes.] is modified by another process already, nested exception message: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update user_table_row set metadata_location=?, storage_type=?, version=? where database_id=? and table_id=? and version=?; nested exception is org.hibernate.StaleStateException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update user_table_row set metadata_location=?, storage_type=?, version=? where database_id=? and table_id=? and version=?","stacktrace":null,"cause":"Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update user_table_row set metadata_location=?, storage_type=?, version=? where database_id=? and table_id=?* Connection #0 to host localhost left intact
 and version=?; nested exception is org.hibernate.StaleStateException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update user_table_row set metadata_location=?, storage_type=?, version=? where database_id=? and table_id=? and version=?"}
```

**Ex4. Put on a deleted entry**
Succeeded. Put will be blindly accepted.

**We can see that there are 2 problems:**
1. Concurrent post should fail, but not taken as a put.
2. Put on a deleted entry should also fail.

**We can also learn an important fact:**
Concurrent puts and concurrent put/delete can be taken care at the repository level. We don't need additional locks or transactions to make the save() method safe. So we only need to tackle the 2 problems at the service level, more specifically in the mapper.

As a matter of fact, in JPA implementation, the save() method will consider an entity as new only if the versioned column is null. If so, it will throw an error on the second post. Therefore, `UserTableVersionMapper` should return null as the version for the new entry instead of `1L`. With this knowledge in mind, we can also understand why concurrent puts or concurrent put/delete can be taken care. The JPA module takes the entry as being seen, but it can not find it as it has been modified or deleted, so it throws an error.

Source: https://docs.spring.io/spring-data/jpa/docs/2.6.1/reference/html/#jpa.entity-persistence

### After change
**Ex1. concurrent posts**
Failed. Second post causes internal server error.
```
{"status":"CONFLICT","error":"Conflict","message":"Entity with key[databaseId : d1, tableId : t1, version: 0 The requested user table has been modified/created by other processes.] is modified by another process already, nested exception message: could not execute statement; SQL [n/a]; constraint [user_table_row.PRIMARY]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute statement","stacktrace":null,"cause":"could not execute statement; SQL [n/a]; constraint [user_table_row.PRIMARY]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute statement"}
```

**Ex2. concurrent puts**
Failed. Second put will conflict.

**Ex3. concurrent put and delete**
Failed. Delete goes before save(), and so put will conflict.

**Ex4. Put on a deleted entry**
Failed. Put will be rejected as the entry is deleted.
```
{"status":"CONFLICT","error":"Conflict","message":"Entity with key[databaseId : d1, tableId : t1 The requested user table has been deleted by other processes.] is modified by another process already, nested exception message: null","stacktrace":null,"cause":null}
```

# Additional Information

- [ ] Breaking Changes
- [ ] Deprecations
- [ ] Large PR broken into smaller PRs, and PR plan linked in the description.

### CODE REVIEW

File: services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java [hunk 1]
```diff
@@ -15,11 +15,21 @@
...
+      if (!userTable.getTableVersion().equals(INITIAL_VERSION)) {
+        throw new EntityConcurrentModificationException(
+            String.format(
+                "databaseId : %s, tableId : %s %s",
+                userTable.getDatabaseId(),
+                userTable.getTableId(),
+                "The requested user table has been deleted by other processes."),
+            new RuntimeException());
+      }
+      return null;
```
Comment (HotSushi) [inline services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java:32]: Whats the long value that gets persisted for a new entry?

File: services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java [hunk 2]
```diff
@@ -15,11 +15,21 @@
  */
 @Mapper(componentModel = "spring")
 public class UserTableVersionMapper {
+  public static final String INITIAL_VERSION = "INITIAL_VERSION";
```
Comment (HotSushi) [inline services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java:18]: lets use `com.linkedin.openhouse.common.api.validator.INITIAL_TABLE_VERSION` like used in `OpenHouseInternalRepositoryImpl`

File: services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java [hunk 3]
```diff
@@ -15,11 +15,21 @@
  */
 @Mapper(componentModel = "spring")
 public class UserTableVersionMapper {
```
Comment (HotSushi) [inline services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java:17]: Can you add the documentation to UserTable APISpec?

```
@Schema(description = "Current Version of the user table.", example = "")
  @JsonProperty(value = "tableVersion")
  private String tableVersion;

```

^ this should include info on `new record should have INTITAL_VERISON`

**Review by HotSushi — CHANGES_REQUESTED** [review]
few comments, but looks great. Great discovery on `As a matter of fact, in JPA implementation, the save() method will consider an entity as new only if the versioned column is null`

File: services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java (hunk 1 above)
Comment (jiang95-dev) [reply services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java:32]: 0

File: services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java (hunk 2 above)
Comment (jiang95-dev) [reply services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java:18]: ok.

File: services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java (hunk 3 above)
Comment (jiang95-dev) [reply services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java:17]: ok.

File: services/housetables/src/main/java/com/linkedin/openhouse/housetables/services/UserTablesServiceImpl.java [hunk 4]
```diff
@@ -68,7 +69,9 @@ public Pair<UserTableDto, Boolean> putUserTable(UserTable userTable) {
 
     try {
       returnedDto = userTablesMapper.toUserTableDto(htsJdbcRepository.save(targetUserTableRow));
-    } catch (CommitFailedException | ObjectOptimisticLockingFailureException ce) {
+    } catch (CommitFailedException
+        | ObjectOptimisticLockingFailureException
+        | DataIntegrityViolationException e) {
```
Comment (autumnust) [inline services/housetables/src/main/java/com/linkedin/openhouse/housetables/services/UserTablesServiceImpl.java:74]: how is this exception encountered and what kind of scenario lead to it ?

File: services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java [hunk 5]
```diff
@@ -15,11 +16,19 @@
...
-
   @Named("toVersion")
   public Long toVersion(UserTable userTable, @Context Optional<UserTableRow> existingUserTableRow) {
```
[the remark this answers is not in the document]
Comment (jiang95-dev) [reply services/housetables/src/main/java/com/linkedin/openhouse/housetables/dto/mapper/UserTableVersionMapper.java:20]: Yes. `UserTableVersionMapperTest::testToVersionWithNoExistingRow` serves the purpose.

File: services/housetables/src/test/java/com/linkedin/openhouse/housetables/e2e/usertable/HtsRepositoryTest.java [hunk 6]
```diff
@@ -58,9 +59,15 @@ public void testDeleteUserTable() {
   @Test
   public void testSaveUserTableWithConflict() {
```
[the remark this answers is not in the document]
Comment (jiang95-dev) [reply services/housetables/src/test/java/com/linkedin/openhouse/housetables/e2e/usertable/HtsRepositoryTest.java:68]: Added one test.

**Review by HotSushi — APPROVED** [review]
great discovery on "null" storage for new row entry. nice clean PR 👏

**Pull request merged**